my_pyscf/mcscf/lasci.py

```diff
@@ -1043,6 +1045,101 @@ def make_casdm2_sub (self, ci=None, ncas_sub=None, nelecas_sub=None, casdm2fr=No
+        casdm2s = []
+        for fcibox, ci_i, ncas, nelecas in zip (self.fciboxes, ci, ncas_sub, nelecas_sub):
+            if ci_i is None:
+                dm2aa = dm2ab, dm2bb = np.zeros ((ncas, ncas,ncas,ncas))
+            else:
+                dm2aa, dm2ab, dm2bb = fcibox.states_make_rdm12s (ci_i, ncas, nelecas)[-1]
+            casdm2s.append (np.stack ([dm2aa, dm2ab, dm2bb], axis=1))
+        return casdm2s
+    
+    #SV states_make_casdm2s
+    def states_make_casdm2s (self, ci=None, ncas_sub=None, nelecas_sub=None,
+            casdm1frs=None, casdm2fr=None, casdm2frs=None, **kwargs):
+        ''' Make the full-dimensional casdm2s spanning the collective active space '''
+        if ci is None: ci = self.ci
+        if ncas_sub is None: ncas_sub = self.ncas_sub
+        if nelecas_sub is None: nelecas_sub = self.nelecas_sub
+        if casdm1frs is None: casdm1frs = self.states_make_casdm1s_sub (ci=ci)
+        if casdm2fr is None: casdm2fr = self.states_make_casdm2_sub (ci=ci,
+            ncas_sub=ncas_sub, nelecas_sub=nelecas_sub, **kwargs)
+        if casdm2frs is None: casdm2frs = self.states_make_casdm2s_sub (ci=ci,
+            ncas_sub=ncas_sub, nelecas_sub=nelecas_sub, **kwargs)
+    
+        ncas = sum (ncas_sub)
+        ncas_cum = np.cumsum ([0] + ncas_sub.tolist ())
+
+        casdm2rs = np.zeros ((self.nroots,3,ncas,ncas,ncas,ncas))
+        for isub, dm2 in enumerate (casdm2frs):
+        #    print ("SV dm2 = ", dm2, dm2.shape, isub, dm2[:,1]) # dm2 = root,spin,ncas,ncas,ncas,ncas | dm2[:,x] = x spin part of ncas,ncas,ncas,ncas
+            i = ncas_cum[isub]
+            j = ncas_cum[isub+1]
+            for spin in [0,1,2]:#0=aa, 1=ab, 2=bb
+                casdm2rs[:, spin, i:j, i:j, i:j, i:j] = dm2[:,spin,:,:,:,:]
+
+        for (isub1, dm1s1_r), (isub2, dm1s2_r) in combinations (enumerate (casdm1frs), 2):
+            i = ncas_cum[isub1]
+            j = ncas_cum[isub1+1]
+            k = ncas_cum[isub2]
+            l = ncas_cum[isub2+1]
+            for spin in [0,1,2]:
+                for dm1s1, dm1s2, casdm2 in zip (dm1s1_r, dm1s2_r, casdm2rs[:,spin]):
+                    dma1, dmb1 = dm1s1[0], dm1s1[1]
+                    dma2, dmb2 = dm1s2[0], dm1s2[1]
+                    # Coulomb slice
+                    casdm2[i:j, i:j, k:l, k:l] = np.multiply.outer (dma1+dmb1, dma2+dmb2)
+                    casdm2[k:l, k:l, i:j, i:j] = casdm2[i:j, i:j, k:l, k:l].transpose (2,3,0,1)
+                    # Exchange slice
+                    casdm2[i:j, k:l, k:l, i:j] = -(np.multiply.outer (dma1, dma2)
+                                               +np.multiply.outer (dmb1, dmb2)).transpose (0,3,2,1)
+                    casdm2[k:l, i:j, i:j, k:l] = casdm2[i:j, k:l, k:l, i:j].transpose (1,0,3,2)
+
+        return casdm2rs
+    
+    #SV make_casdm3_sub
+    def make_casdm3_sub (self, ci=None, ncas_sub=None, nelecas_sub=None, casdm3fr=None, **kwargs):
```

Comment:
Why is this function necessary?

Reply:
I didn't use it in make_casdm3 but while understanding the workflow, I just made it. It is not needed. I have removed it.